src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -216,11 +214,26 @@ toSnapIncomingRun (Merging mergePolicy mergingRun) = do
         (SuppliedCredits suppliedCredits)
         smrs
 
+-- | Only to be used for incoming runs! Merging runs inside a merging tree need
+-- to use MR.TreeMergeType.
+--
+-- We could offer a more general type if 'Mr.MergingRunState' also had a type
+-- parameter 't', but that adds quite a bit of noise.
 toSnapMergingRunState ::
      MR.MergingRunState m h
-  -> SnapMergingRunState (Ref (Run m h))
+  -> SnapMergingRunState MR.LevelMergeType (Ref (Run m h))
 toSnapMergingRunState (MR.CompletedMerge r)  = SnapCompletedMerge r
-toSnapMergingRunState (MR.OngoingMerge rs m) = SnapOngoingMerge rs (Merge.mergeType m)
+toSnapMergingRunState (MR.OngoingMerge rs m) =
+    -- The merge type conversion should never fail, since the Merge was
+    -- constructed based on the restricted LevelMergeType.
+    -- To make the types align, we could either:
+    -- * store the restricted type inside MergingRun itself, but this would
+    --   just duplicate information, not provide any safety.
+    -- * Also add the type parameter to Merge, adding complexity there.
```

Comment:
Does it really make the code more complex, or is it a noisy change? If it's mostly noise, I think we should/could go for it

Reply:
It's mainly noisy, and we also get more indirection in `Merge.steps` (an additional type class dictionary call for every entry we merge). That's probably okay, but not sure how it impacts other optimisations.

Reply:
You can precalculate `t -> MergeType` and store the result somewhere, no need to compute it on every call to `Merge.steps`

Reply:
I'll merge this PR as it is now and open a new one for propagating the type parameter further inwards. Then we can discuss that separately.